### ISSUE TITLE
add support for languages with composition input

### ISSUE THREAD

**Issue by ianstormtaylor** [ianstormtaylor/slate#74]
Right now none of the `composition` events are handled, so non-latin languages that require this won't work correctly.

**Comment by hansanghoon** [comment]
Hello. I recorded some video typing Korean Characters in Auto-Markdown mode. Is this the situation where non-latin languages doesn't work properly?

[slatejs_koreanchar.zip](https://github.com/ianstormtaylor/slate/files/366515/slatejs_koreanchar.zip)

**Comment by ianstormtaylor** [comment]
Ah yup, that looks like it! We need to add support for the [`composition` events](https://developer.mozilla.org/en-US/docs/Web/API/CompositionEvent) that are triggered when using Korean and other languages.

**Comment by ianstormtaylor** [comment]
Hey @hansanghoon I've just added the most basic composition support in `0.2.2`, could you test it out again and see if that works for you? Thanks!

**Comment by hansanghoon** [comment]
[160719_slate_rich.zip](https://github.com/ianstormtaylor/slate/files/370497/160719_slate_rich.zip)
Auto-markdown still broke with exception so I would not upload its recorded video. Rich Text works nicer than before, but still have some problem. Typing in formal sequence looks fine, but if I combine typing with arrow key, slate put caret in wrong place. 

which part of slatejs code handle this situation(composition events?)? I'm not good at coding, but could do some...thing like trial-and-error testing.

**Comment by ianstormtaylor** [comment]
@hansanghoon good catch. I've just published `0.2.4` which should fix this problem. It now ends the composition when any of those keys (arrow, enter, etc.) are pressed instead of messing up. Is this the behavior you'd expect? is it similar to the behavior of other editors a user would have used?

**Comment by ianstormtaylor** [comment]
(Make that `0.2.5`, since I discovered and issue with `0.2.4`!)

**Comment by hansanghoon** [comment]
@ianstormtaylor Auto markdown headings with Korean character now works fine :)

[160720_compositionandarrow.zip](https://github.com/ianstormtaylor/slate/files/372723/160720_compositionandarrow.zip)
For composition input with arrow key part. I made comparison videos. slate vs note(mac app). During composition caret is placed on right side of character. when I press left key, composition ends and caret moves to left side of a character. when I press right key, composition ends and caret jumps over right side character. What we saw in a video of apple note app is typical behaviour. 

[160720_slate_cons.zip](https://github.com/ianstormtaylor/slate/files/372742/160720_slate_cons.zip)
consecutive composition input and right arrow. If I press right key twice, new composition take place in right place. If I press right key only once, second composition starts and new composed chars are moved to left by on char length.

Maybe slate needs more extensive test cases. Hope this helps!

**Comment by ianstormtaylor** [comment]
@hansanghoon thanks again for helping me track this down. I agree we need more test cases for Slate. Workings on a few more today for transforms and other non-browser pieces, but the browser ones can be really hard. Do you have any tips for testing IME-type interactions?

The update: I've pushed `0.5.1` which fixes the "consecutive compositions doing weird things" bug, so that hopefully writing now doesn't break at all.

The downside is that I haven't gotten it working as seamless as macOS's Note app. I've looked at Draft.js and they also have this "non-seamless" behavior, so I'm going to call it good enough for now since the writing itself is possible, just the arrow keys "end" instead of "end and move". I hope that works for you!

If you interested (or anyone else), I'd be down for a pull request that solves the arrow key issues. And if you run into any other troubles just let me know (on Slack or in another issue)!
